contracts/utils/ComplementaryServiceMetadata.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.30;
+
+/// @dev Service Registry interface
+interface IServiceRegistry {
+    enum ServiceState {
+        NonExistent,
+        PreRegistration,
+        ActiveRegistration,
+        FinishedRegistration,
+        Deployed,
+        TerminatedBonded
+    }
+
+    /// @dev Gets the service instance from the map of services.
+    /// @param serviceId Service Id.
+    /// @return securityDeposit Registration activation deposit.
+    /// @return multisig Service multisig address.
+    /// @return configHash IPFS hash pointing to the config metadata.
+    /// @return threshold Agent instance signers threshold.
+    /// @return maxNumAgentInstances Total number of agent instances.
+    /// @return numAgentInstances Actual number of agent instances.
+    /// @return state Service state.
+    function mapServices(uint256 serviceId) external view returns (uint96 securityDeposit, address multisig,
+        bytes32 configHash, uint32 threshold, uint32 maxNumAgentInstances, uint32 numAgentInstances, ServiceState state);
+
+    /// @dev Gets the owner of a specified service Id.
+    /// @param serviceId Service Id.
+    /// @return serviceOwner Service owner address.
+    function ownerOf(uint256 serviceId) external view returns (address serviceOwner);
+}
+
+/// @dev Provided zero address.
+error ZeroAddress();
+
+/// @dev Caught reentrancy violation.
+error ReentrancyGuard();
+
+/// @dev Account is unauthorized.
+/// @param account Account address.
+error UnauthorizedAccount(address account);
+
+/// @title ComplementaryServiceMetadata - Complementary Service Metadata contract to manage complementary service hashes
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+/// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
+contract ComplementaryServiceMetadata {
+    event ComplementaryMetadataUpdated(uint256 indexed serviceId, bytes32 indexed hash);
+
+    // Olas mech version number
+    string public constant VERSION = "0.1.0";
+
+    // Service Registry address
+    address public immutable serviceRegistry;
+
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    // Mapping of serviceId => actual metadata hash
+    mapping(uint256 => bytes32) public mapServiceHashes;
```

Comment:
imo we should apply the same sanitising for the ipfs string pattern we do in the service registry

Reply:
We don't do anything specific there except for returning the URI with the prefix Do we want to use that here? And use `baseURI` from serviceRegistry?

Reply:
Yes let's return the full url composed from prefix etc